Flaky-test: TransactionMetaStoreAssignmentTest.testTransactionMetaStoreAssignAndFailover

<!--- 

Instructions for reporting a flaky test using this issue template:

1. Replace [test class] in title and body with the test class name
2. Replace [test method] in title and body with the test method that failed. Multiple methods are flaky, remove the content that refers to the test method.
3. Replace "url here" with a url to an example failure. In the Github Actions workflow run logs, you can right click on the line number to copy a link to the line. Example of such url is https://github.com/apache/pulsar/pull/8892/checks?check_run_id=1531075794#step:9:377 . The logs are available for a limited amount of time (usually for a few weeks).
4. Replace "relevant parts of the exception stacktrace here" with the a few lines of the stack trace that shows at leat the exception message and the line of test code where the stacktrace occured.
5. Replace "full exception stacktrace here" with the full exception stacktrace from logs. This section will be hidded by default.
6. Remove all unused fields / content to unclutter the reported issue. Remove this comment too.

-->
TransactionMetaStoreAssignmentTest is flaky. The testTransactionMetaStoreAssignAndFailover test method fails sporadically.

[example failure](https://github.com/apache/pulsar/pull/10782/checks?check_run_id=2728428471#step:10:1649)

```
testTransactionMetaStoreAssignAndFailover(org.apache.pulsar.broker.transaction.coordinator.TransactionMetaStoreAssignmentTest)  Time elapsed: 0.008 s  <<< FAILURE!
java.lang.AssertionError: expected [16] but found [9]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.broker.transaction.coordinator.TransactionMetaStoreAssignmentTest.testTransactionMetaStoreAssignAndFailover(TransactionMetaStoreAssignmentTest.java:40)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

[another failure](https://github.com/apache/pulsar/pull/10782/checks?check_run_id=2728428471#step:10:925)
```
testTransactionMetaStoreAssignAndFailover(org.apache.pulsar.broker.transaction.coordinator.TransactionMetaStoreAssignmentTest)  Time elapsed: 0.007 s  <<< FAILURE!
java.lang.AssertionError: expected [16] but found [12]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.broker.transaction.coordinator.TransactionMetaStoreAssignmentTest.testTransactionMetaStoreAssignAndFailover(TransactionMetaStoreAssignmentTest.java:40)
```

[recent failure](https://github.com/apache/pulsar/pull/11046/checks?check_run_id=2894997728#step:10:2128):
```
Error:  Tests run: 4, Failures: 1, Errors: 0, Skipped: 3, Time elapsed: 11.741 s <<< FAILURE! - in org.apache.pulsar.broker.transaction.coordinator.TransactionMetaStoreAssignmentTest
Error:  testTransactionMetaStoreAssignAndFailover(org.apache.pulsar.broker.transaction.coordinator.TransactionMetaStoreAssignmentTest)  Time elapsed: 0.017 s  <<< FAILURE!
java.lang.AssertionError: expected [16] but found [13]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.broker.transaction.coordinator.TransactionMetaStoreAssignmentTest.testTransactionMetaStoreAssignAndFailover(TransactionMetaStoreAssignmentTest.java:40)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

@Renkai @congbobo184 @gaoran10 Would you be able to check what's causing the flakiness in this test? thank you

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/coordinator/TransactionMetaStoreAssignmentTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/coordinator/TransactionMetaStoreAssignmentTest.java
@@ -20,6 +20,7 @@ package org.apache.pulsar.broker.transaction.coordinator;
 
 import java.io.IOException;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import org.apache.pulsar.broker.PulsarService;
 import org.awaitility.Awaitility;
@@ -32,12 +33,13 @@ public class TransactionMetaStoreAssignmentTest extends TransactionMetaStoreTest
     @Test(groups = "broker")
     public void testTransactionMetaStoreAssignAndFailover() throws IOException {
 
-        int transactionMetaStoreCount = 0;
-        for (PulsarService pulsarService : pulsarServices) {
-            transactionMetaStoreCount += pulsarService.getTransactionMetadataStoreService().getStores().size();
-        }
-
-        Assert.assertEquals(transactionMetaStoreCount, 16);
+        Awaitility.await()
+                .untilAsserted(() -> {
+                    int transactionMetaStoreCount = Arrays.stream(pulsarServices)
+                            .mapToInt(pulsarService -> pulsarService.getTransactionMetadataStoreService().getStores().size())
+                            .sum();
+                    Assert.assertEquals(transactionMetaStoreCount, 16);
+                });
 
         PulsarService crashedMetaStore = null;
         for (int i = pulsarServices.length - 1; i >= 0; i--) {
@@ -62,16 +64,11 @@ public class TransactionMetaStoreAssignmentTest extends TransactionMetaStoreTest
 
         Awaitility.await()
                 .untilAsserted(() -> {
-
-                    int transactionMetaStoreCount2 = 0;
-                    for (PulsarService pulsarService : pulsarServices) {
-                        transactionMetaStoreCount2 += pulsarService.getTransactionMetadataStoreService().getStores()
-                                .size();
-                    }
-
+                    int transactionMetaStoreCount2 = Arrays.stream(pulsarServices)
+                            .mapToInt(pulsarService -> pulsarService.getTransactionMetadataStoreService().getStores().size())
+                            .sum();
                     Assert.assertEquals(transactionMetaStoreCount2, 16);
                 });
-
         transactionCoordinatorClient.close();
     }
 }

